contracts/managers/context/ContextManager.sol

```diff
@@ -0,0 +1,201 @@
+        uint256 length = executeData.addrs.length;
+        // Check arrays length match
+        if (executeData.data.length != length) {
+            revert LengthMismatch();
+        }
+
+        results = new bytes[](length);
+
+        address target;
+        bytes calldata data;
+        for (uint256 i; i < length; ++i) {
+            target = executeData.addrs[i];
+            data = executeData.data[i];
+
+            // Check if address is approved
+            if (!ContextManagerStorageLib.isApproved(target)) {
+                revert AddressNotApproved(target);
+            }
+
+            IContextClient(target).setupContext(msg.sender);
+
+            // Execute call
+            results[i] = target.functionCall(data);
+
+            IContextClient(target).clearContext();
+
+            emit ContextCall(target, data, results[i]);
+        }
+    }
+
+    /// @notice Executes multiple calls to approved addresses with signature verification
+    /// @param contextDataArray Array of context data containing signatures and call data
+    /// @return results Array of results from each call
+    function runWithContextAndSignature(
+        ContextDataWithSender[] calldata contextDataArray
+    ) external returns (bytes[] memory results) {
+        uint256 length = contextDataArray.length;
+        results = new bytes[](length);
+
+        ContextDataWithSender calldata contextData;
+        for (uint256 i; i < length; ++i) {
+            contextData = contextDataArray[i];
+
+            // Check if signature has expired
+            if (block.timestamp > contextData.expirationTime) {
+                revert SignatureExpired();
+            }
+
+            // Verify signature
+            if (!_verifySignature(contextData)) {
+                revert InvalidSignature();
+            }
+
+            // Verify and update nonce
+            ContextManagerStorageLib.verifyAndUpdateNonce(contextData.sender, contextData.nonce);
+
+            // Check if target address is approved
+            if (!ContextManagerStorageLib.isApproved(contextData.target)) {
+                revert AddressNotApproved(contextData.target);
+            }
+
+            IContextClient(contextData.target).setupContext(contextData.sender);
+
+            // Execute call
+            results[i] = contextData.target.functionCall(contextData.data);
+
+            IContextClient(contextData.target).clearContext();
+
+            emit ContextCall(contextData.target, contextData.data, results[i]);
+        }
+    }
+
+    function isApproved(address addr) external view returns (bool) {
+        return ContextManagerStorageLib.isApproved(addr);
+    }
+
+    function getApprovedAddressesList() external view returns (address[] memory) {
+        return ContextManagerStorageLib.getApprovedAddressesList();
+    }
+
+    function _verifySignature(ContextDataWithSender memory contextData) internal pure returns (bool) {
```

Comment:
Verify if this implementation is resistant to replay attack. Check if you are follow witn EIP-712 standards like chainId should be a part of signature

Reply:
We should be prepared for multichains

Reply:
In a "replay attack," a nonce is applied, so we are resistant to it.